pkg/reconciler/node/node.go

```diff
@@ -0,0 +1,53 @@
+package node
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/tektoncd/pipeline/pkg/workspace"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	nodereconciler "knative.dev/pkg/client/injection/kube/reconciler/core/v1/node"
+	"knative.dev/pkg/logging"
+	pkgreconciler "knative.dev/pkg/reconciler"
+)
+
+// Reconciler implements controller.Reconciler for Configuration resources.
+type Reconciler struct {
+	KubeClientSet kubernetes.Interface
+}
+
+// Check that our Reconciler implements noderunreconciler.Interface
+var (
+	_ nodereconciler.Interface = (*Reconciler)(nil)
+)
+
+// ReconcileKind deletes any affinity assistant pods on unschedulable nodes
+func (c *Reconciler) ReconcileKind(ctx context.Context, n *corev1.Node) pkgreconciler.Event {
+	if !n.Spec.Unschedulable {
+		return nil
+	}
+	logger := logging.FromContext(ctx)
+	logger.Debugf("found unschedulable node %s", n.Name)
+
+	// Find all affinity assistant pods on the node
+	pods, err := c.KubeClientSet.CoreV1().Pods("").List(ctx, metav1.ListOptions{
+		LabelSelector: workspace.LabelComponent + "=" + workspace.ComponentNameAffinityAssistant,
+		FieldSelector: "spec.nodeName=" + n.Name,
+	})
+	if err != nil {
+		return fmt.Errorf("couldn't list affinity assistant pods on node %s: %s", n.Name, err)
+	}
+
+	// Delete affinity assistant pods, allowing their statefulset controllers to recreate them.
```

Comment:
thanks @lbernick 🙏 this is very exciting

While I am trying to understand this, what happens when the existing taskRuns are not completed but the associated AA pod is gone?

Reply:
Existing taskruns should run to completion as long as the pods are not deleted, since the affinity terms apply only while the pod is being scheduled, not while it's running.

Reply:
thanks @lbernick 

Can we scope this kind of deletion to a pipelineRun instead of deleting all AA pods on an unscheduable node? No specific reasons as such for this kind of request, just evaluating isolation or any unintentional corruption.

Reply:
I'm not sure what you mean exactly-- it sounds like you want to delete and recreate the affinity assistant pod as part of the pipelinerun reconciler loop? If we wanted to do it that way, we could add a check in the reconcile loop where we get the affinity assistant pod, see what node it's running on, and see if that node is unschedulable. However, I'm not sure how we'd trigger a reconcile loop when a node becomes unschedulable, which is why I chose to do it this way.

Reply:
> it sounds like you want to delete and recreate the affinity assistant pod as part of the pipelinerun reconciler loop?

I would like to delete the affinity assistant pod which belongs to my pipeline instead of deleting my neighbors as well. I think we can add a label selector to match with the pipelineRun name while deleting the pod if we can somehow access the pipelineRun name in the node controller.

Reply:
Or instead of deleting affinity assistant pod as soon as the node is cordoned or unschedulable, delay deletion until the next task is scheduled. So before creating a pod/taskRun for the next scheduled task, check the node where affinity assistant is located. If that node is unschedulable, delete existing affinity assistant pod. I am assuming the further processing of creating affinity assistant pod followed by the taskRun pod stays the same.

Reply:
Update: Priti is exploring the alternative approach in https://github.com/tektoncd/pipeline/issues/6586

I'm a bit concerned the approach suggested here could lead to race conditions. For example, even if we check before creating a TaskRun that the affinity assistant node is schedulable, it may become unschedulable before the taskrun pod can be scheduled.